src/Repository/DeviceCodeRepository.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+declare(strict_types=1);
+
+namespace League\Bundle\OAuth2ServerBundle\Repository;
+
+use League\Bundle\OAuth2ServerBundle\Converter\ScopeConverterInterface;
+use League\Bundle\OAuth2ServerBundle\Entity\DeviceCode as DeviceCodeEntity;
+use League\Bundle\OAuth2ServerBundle\Manager\ClientManagerInterface;
+use League\Bundle\OAuth2ServerBundle\Manager\DeviceCodeManagerInterface;
+use League\Bundle\OAuth2ServerBundle\Model\AbstractClient;
+use League\Bundle\OAuth2ServerBundle\Model\DeviceCode as DeviceCodeModel;
+use League\Bundle\OAuth2ServerBundle\Model\DeviceCodeInterface;
+use League\OAuth2\Server\Entities\DeviceCodeEntityInterface;
+use League\OAuth2\Server\Exception\OAuthServerException;
+use League\OAuth2\Server\Repositories\ClientRepositoryInterface;
+use League\OAuth2\Server\Repositories\DeviceCodeRepositoryInterface;
+
+final class DeviceCodeRepository implements DeviceCodeRepositoryInterface
+{
+    /**
+     * @var DeviceCodeManagerInterface
+     */
+    private $deviceCodeManager;
+
+    /**
+     * @var ClientManagerInterface
+     */
+    private $clientManager;
+
+    /**
+     * @var ScopeConverterInterface
+     */
+    private $scopeConverter;
+
+    /**
+     * @var ClientRepositoryInterface
+     */
+    private $clientRepository;
+
+    public function __construct(
+        DeviceCodeManagerInterface $deviceCodeManager,
+        ClientManagerInterface $clientManager,
+        ScopeConverterInterface $scopeConverter,
+        ClientRepositoryInterface $clientRepository,
+    ) {
+        $this->deviceCodeManager = $deviceCodeManager;
+        $this->clientManager = $clientManager;
+        $this->scopeConverter = $scopeConverter;
+        $this->clientRepository = $clientRepository;
+    }
+
+    public function getNewDeviceCode(): DeviceCodeEntityInterface
+    {
+        return new DeviceCodeEntity();
+    }
+
+    public function persistDeviceCode(DeviceCodeEntityInterface $deviceCodeEntity): void
+    {
+        $deviceCode = $this->deviceCodeManager->find($deviceCodeEntity->getIdentifier());
+        $newDeviceCode = false;
+
+        if ($deviceCode) {
+            if ($deviceCodeEntity->getLastPolledAt()) {
+                $deviceCode->setLastPolledAt($deviceCodeEntity->getLastPolledAt());
+            }
+        } else {
+            $newDeviceCode = true;
+            $deviceCode = $this->buildDeviceCodeModel($deviceCodeEntity);
+        }
+
+        $this->deviceCodeManager->save($deviceCode, $newDeviceCode);
```

Comment:
I think there is no need to update the `lastPolledAt` property. `DeviceGrant` is already doing it

I think this method should build a new `DeviceCode` model if not exists (`buildDeviceCodeModel`), or update it  (I would add a `updateDeviceCodeModel` method) with the values of the received `DeviceCodeEntityInterface` and finally save it to the `DeviceManagerInteface`